### PR TITLE
fix(web): Let embedded device handle K_TAB or K_ENTER

### PR DESCRIPTION
**Pull request by darcywong00**
I noticed `K_TAB` and `K_ENTER` aren't working on Keyman for Android with an external keyboard (again 😞 )

KMEA is expecting `kmw['executeHardwareKeystroke']` to return true if KMW didn't process certain keys 
(do we also include `K_TABBACK`, `K_TABFWD`)?

### CODE REVIEW

**Comment by darcywong00** [comment]
hmm, although maybe KMW still needs to pass `K_ENTER` to the lmlayer... ?

**Comment by mcdurdin** [comment]
I think I'd like to understand how this broke -- does @jahorton have any thoughts here?

**Comment by jahorton** [comment]
> I think I'd like to understand how this broke -- does @jahorton have any thoughts here?

Depends... when was the last 'fix', again?

**Comment by darcywong00** [comment]
> Depends... when was the last 'fix', again?

Going by the history, it was
> ## 2020-01-31 13.0.6051 beta
>  * Bug fix:
>     * Fix external keyboard key "tab" (updated Keyman Web Engine #2546)

**Review by jahorton — COMMENTED** [review]
OK.  If the last change was that far back, it'd have been before the web-core refactor.

Said refactor did involve some changes to how keys are handled by default, so it's likely that TAB and ENTER previously weren't handled but now _are_ by Web's core keystroke processing.  Bypassing this sounds like the right call for most cases.

I am a bit concerned that this may affect the context somewhat, though.  Typing within a mail app or Google Docs still needs to process an actual `\n` from ENTER keystrokes, to name a few cases.  We may need some way to signal the 'input mode' - which behavior the keys represent at a given moment - so that those cases don't break.

**Comment by mcdurdin** [comment]
> I am a bit concerned that this may affect the context somewhat, though. Typing within a mail app or Google Docs still needs to process an actual `\n` from ENTER keystrokes, to name a few cases. We may need some way to signal the 'input mode' - which behavior the keys represent at a given moment - so that those cases don't break.

If we fallback to the default behaviour for <kbd>Enter</kbd> and <kbd>Tab</kbd>, and a multiline edit control inserts a `\n`, then that should trigger a context change anyway, because the underlying content will have been changed, right?

**Comment by jahorton** [comment]
> If we fallback to the default behaviour for Enter and Tab, and a multiline edit control inserts a `\n`, then that should trigger a context change anyway, because the underlying content will have been changed, right?

Probably depends on the control.  A 'textedit' control on a webpage would simply append a `\n` to its `.text` value, not throw out the whole context.  If a user then hit backspace to remove the `\n`, the old text should still be there.  Of course, if the `\n` _isn't_ there (because the keystroke processor never saw it), the backspace might delete the wrong character.

**Comment by mcdurdin** [comment]
> unless the Android app takes steps to re-sync Web's context data.

This is the job of Keyman Engine, isn't it?

**Comment by jahorton** [comment]
> > unless the Android app takes steps to re-sync Web's context data.
> 
> This is the job of Keyman Engine, isn't it?

Keyman Engine for Web, or Keyman Engine for Android?  Both parts we're talking about are technically "Keyman Engine."

**Comment by mcdurdin** [comment]
> Keyman Engine for Web, or Keyman Engine for Android? Both parts we're talking about are technically "Keyman Engine."

Keyman Engine for Android should be capturing external changes to the context (whether that's by change of focus, apps inserting/modifying content, or through secondary inputs such as voice, key events that are passed through, or anything else), and passing those context changes in to Keyman Engine for Web.

**Comment by jahorton** [comment]
> Keyman Engine for Android should be capturing external changes to the context (whether that's by change of focus, apps inserting/modifying content, or through secondary inputs such as voice, key events that are passed through, or anything else), and passing those context changes in to Keyman Engine for Web.

OK, let's work with that for a second.  So, what about when keystrokes should effect external changes?  Keep in mind that TAB and ENTER tend to be overloaded keys.

If a user is physically giving input to a word processor, TAB and ENTER are most likely intended as text output keys, resulting in mutations to the existing context.  I'm going to try to make things very explicit here - the "text source"/"input field" remains the same, though the "context" has changed.

In other scenarios, like when navigating UI, a form, or common webpage use, TAB and ENTER usually trigger UI events.  There is no existing "text source" / "input field"... and if there is, often these keystrokes trigger a change in which one the engine is working with.  It's not just a change of context, but of text source.  These are fundamentally different.

Keyman Engine for Web _cannot_ know which of these two states Keyman Engine for Android is in without additional information that only Keyman Engine for Android can access.  Despite this, we do expect some differences in behavior:

Predictions:

- should be completely reset on a change of text source, also invalidating the existing text context within KMW.
- should continue, without invalidating text context within KMW.

**Comment by mcdurdin** [comment]
> Keyman Engine for Web _cannot_ know which of these two states Keyman Engine for Android is in without additional information that only Keyman Engine for Android can access. Despite this, we do expect some differences in behavior:
> 
> Predictions:
> 
> * should be completely reset on a change of text source, also invalidating the existing text context within KMW.
> * should continue, without invalidating text context within KMW.

Correct.

There is another effect, which is when the app modifies the text store for the focused edit control directly or indirectly. This is in effect what happens when we pass the <kbd>Enter</kbd> (and some other) keys through to the control, as right now we are giving responsibility for handling that key event to the application (which probably lets it go through to the text edit control, which inserts a `\n`). We may not be able to reliably treat this context change differently to focus changes, etc, at this point. 

I am reasonably okay with treating changes resulting from an unhandled <kbd>Enter</kbd> as a change of context, resetting predictive text and keyboard context. It's not ideal, but we'd really want to make sure we get handling this right. I regard handling this as an optimisation or improvement we can make in a future update.

For now, I just want to make sure that Keyman Engine for Android is capturing external text store changes (whether triggered by <kbd>Enter</kbd> or by gestures or anything else) and passing them through to Keyman Engine for Web.

**Comment by mcdurdin** [comment]
FWIW, this directly relates to the fix in #3662!

**Comment by darcywong00** [comment]
I reverted my original change. On tracking K_ENTER through the Chrome debugger I got to the end of kmembedded.ts
```
return keyman.core.processKeyEvent(Lkc) != null;
```

`keyman.core.processKeyEvent(Lkc)` was null but
`keyman.core.processKeyEvent(Lkc) != null` was `false`

File: web/source/kmwembedded.ts [hunk 1]
```diff
@@ -531,7 +525,7 @@ namespace com.keyman.text {
     try {
       // Now that we've manually constructed a proper keystroke-sourced KeyEvent, pass control
       // off to the processor for its actual execution.
-      return keyman.core.processKeyEvent(Lkc) != null;
+      return !keyman.core.processKeyEvent(Lkc);
```
Comment (jahorton) [inline web/source/kmwembedded.ts:528]: If the original check was for something to not be `null`, you want something truthy.  A single `!` coerces to `false`, so use a double (`!!`) to convert something truthy to `true`.

```suggestion
      return !!keyman.core.processKeyEvent(Lkc);
```

Comment (jahorton) [reply web/source/kmwembedded.ts:528]: I'm guessing what you actually got was `undefined`?  A quick console check does verify that `undefined != null` results in `false`.

That said... I should probably check and see where the `undefined` is coming from.  That sounds unintended.

Comment (jahorton) [reply web/source/kmwembedded.ts]: Looking at the method... it looks like its return types never got fully reworked.  The intent was for the method to return either a `RuleBehavior` or `null` for any case, but there are some `undefined` and `true` cases remaining there.  I should probably address that once my current predictive tasks are done.

**Review by jahorton — CHANGES_REQUESTED** [review]
I'm almost certain this will have unintended side effects as it currently is.  We need to understand the problem - on both sides - more thoroughly before proceeding.

**Comment by mcdurdin** [comment]
> We need to understand the problem - on both sides - more thoroughly before proceeding.

Agreed. Do we want to discuss at Standup today?

**Comment by mcdurdin** [comment]
See also #2558

**Comment by jahorton** [comment]
While looking through the issue list, I noticed https://github.com/keymanapp/keyman/issues/2558.  It might be worth checking for interactions with the issues noted there?

File: web/source/kmwembedded.ts [hunk 2]
```diff
@@ -525,7 +525,7 @@ namespace com.keyman.text {
...
```
Comment (jahorton) [inline web/source/kmwembedded.ts:528]: After a more thorough investigation, LGTM now.  That said... please add a bit of documentation to make this clearer for the future:

```suggestion
      // Should return `false` when Keyman _has_ fully handled the event and `true` when not.
      return !keyman.core.processKeyEvent(Lkc);
```

**Review by jahorton — APPROVED** [review]

File: web/source/kmwembedded.ts (hunk 2 above)
Comment (darcywong00) [reply web/source/kmwembedded.ts:531]: I've added comments about the expected return values.

But still might have some loose ends:
1. While retesting, I noticed K_SPACE also has no effect in the Keyman Text area.
2. After KMW passes K_TAB / K_ENTER back to KMEA to dispatch the key, should I also call KMManager.resetContext() afterwards?

This still wouldn't handle the scenario in #3662 where "Clear Text" updates the KMW text to `""`

Comment (mcdurdin) [reply web/source/kmwembedded.ts:531]: > 2\. After KMW passes K_TAB / K_ENTER back to KMEA to dispatch the key, should I also call KMManager.resetContext() afterwards?

Ideally, that should be resolved by tracking context changes made by the app or other systems.

Comment (jahorton) [reply web/source/kmwembedded.ts:531]: > While retesting, I noticed K_SPACE also has no effect in the Keyman Text area.

After I did my deeper dive into that part of KMW, I kind of expected this.  It's treated very similarly to K_TAB and K_ENTER internally, as many keyboards rely on default behavior here.

Comment (jahorton) [reply web/source/kmwembedded.ts:531]: Just did up https://github.com/keymanapp/keyman/pull/3707 as a patch to address this issue.

**Pull request merged**